README.md
# Python Sudoku App (1.0.0)
![Tests](https://github.com/JohN100x1/sudoku_solver/workflows/Tests/badge.svg?branch=master)
[![Python](https://img.shields.io/badge/python-3.10%2B-brightgreen)](https://www.python.org/)
[![Style](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

A simple Sudoku app made with Tkinter.
- `New Game`: Generates a new Sudoku board.
- `Solve`: Solves the current Sudoku board.
- `Hint`: Fills in one of the answers on the Sudoku board.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-04-15
### Added
- Initial release.

tests/conftest.py
from random import randint


class Factory:
    @staticmethod
    def create_random_board():
        return [[randint(1, 9) for _ in range(9)] for _ in range(9)]

    @staticmethod
    def create_empty_board():
        return [[0 for _ in range(9)] for _ in range(9)]


tests/test_app.py
import sys

import pytest

from sudoku.app import Actions, SudokuApp
from sudoku.view import SudokuBoard


@pytest.fixture
def app():
    return SudokuApp()


@pytest.mark.skipif(sys.platform != "win32", reason="Windows tests")
class TestSudokuAppButtonAction:
    """Test SudokuApp.button_action."""

    def test_generate_board(self, mocker, app):
        mock = mocker.spy(SudokuBoard, "generate")
        app.button_action(Actions.NEW_GAME)
        assert mock.call_count == 1

    def test_solve_board(self, mocker, app):
        mock = mocker.spy(SudokuBoard, "solve")
        app.button_action(Actions.SOLVE)
        assert mock.call_count == 1

    def test_hint_board(self, mocker, app):
        mock = mocker.spy(SudokuBoard, "hint")
        app.button_action(Actions.HINT)
        assert mock.call_count == 1


tests/test_logic.py
from random import seed as random_seed

import pytest
from conftest import Factory

from sudoku.logic import SudokuLogic


@pytest.fixture
def logic():
    return SudokuLogic()


@pytest.fixture
def sample_logic():
    logic = SudokuLogic()
    logic.board = [
        [0, 5, 4, 7, 8, 0, 0, 6, 3],
        [8, 0, 0, 0, 5, 0, 0, 7, 0],
        [7, 0, 0, 4, 0, 1, 2, 5, 0],
        [9, 3, 0, 0, 0, 6, 1, 0, 0],
        [0, 0, 0, 0, 1, 0, 0, 0, 4],
        [4, 1, 0, 3, 0, 0, 0, 0, 0],
        [0, 0, 7, 1, 6, 4, 0, 8, 2],
        [2, 4, 5, 8, 7, 0, 0, 1, 6],
        [6, 0, 0, 2, 0, 0, 7, 0, 9],
    ]
    logic.solution = [
        [1, 5, 4, 7, 8, 2, 9, 6, 3],
        [8, 2, 9, 6, 5, 3, 4, 7, 1],
        [7, 6, 3, 4, 9, 1, 2, 5, 8],
        [9, 3, 8, 5, 4, 6, 1, 2, 7],
        [5, 7, 2, 9, 1, 8, 6, 3, 4],
        [4, 1, 6, 3, 2, 7, 8, 9, 5],
        [3, 9, 7, 1, 6, 4, 5, 8, 2],
        [2, 4, 5, 8, 7, 9, 3, 1, 6],
        [6, 8, 1, 2, 3, 5, 7, 4, 9],
    ]
    return logic


class TestSudokuLogicGetHint:
    """Test SudokuLogic.get_hint."""

    def test_get_hint(self, sample_logic):
        random_seed(1234567809)

        assert sample_logic.get_hint() == (1, 1, 2)


class TestSudokuLogicGenerate:
    """Test SudokuLogic.generate."""

    def test_generate_with_seed(self, logic, sample_logic):
        seed = 1234567809
        num_clues = 40
        keep_pos = logic.generate(seed, num_clues)
        assert len(keep_pos) == num_clues
        assert logic.board == sample_logic.board


class TestSudokuLogicPossible:
    """Test SudokuLogic.possible."""

    def test_possible(self, sample_logic):
        assert sample_logic.possible(1, 0, 0) is True

    def test_col_exclude(self, sample_logic):
        assert sample_logic.possible(9, 0, 0) is False

    def test_row_exclude(self, sample_logic):
        assert sample_logic.possible(3, 0, 0) is False

    def test_box_exclude(self, sample_logic):
        assert sample_logic.possible(3, 4, 0) is False


class TestSudokuLogicIterateBoard:
    """Test SudokuLogic.iterate_board."""

    def test_next_value(self):
        assert next(SudokuLogic.iterate_board()) == (0, 0)


class TestSudokuLogicFindEmpty:
    """Test SudokuLogic.find_empty."""

    def test_found(self, logic):
        assert logic.find_empty() == (0, 0)

    def test_not_found(self, logic):
        for i, j in logic.iterate_board():
            logic.board[i][j] = 1
        assert logic.find_empty() is None


class TestSudokuLogicResetBoard:
    """Test SudokuLogic.reset_board."""

    def test_reset(self, logic):
        logic.board = Factory.create_random_board()
        logic.reset_board()
        assert logic.board == Factory.create_empty_board()


class TestSudokuLogicFillBoard:
    """Test SudokuLogic.fill_board."""

    def test_fill(self, logic):
        solution_board = Factory.create_random_board()
        logic.board = Factory.create_random_board()
        logic.solution = solution_board
        logic.fill_board()
        assert logic.board == solution_board


class TestSudokuLogicSolveBoard:
    """Test SudokuLogic.solve_board."""

    def test_solve(self, sample_logic):
        sample_logic.solve_board()
        assert sample_logic.board == sample_logic.solution


src/sudoku.py
import logging
import random
from tkinter import Button, Entry, LabelFrame, StringVar, Tk

DIGITS = {str(i) for i in range(1, 11)}

logger = logging.getLogger(__name__)


class SudokuApp(Tk):
    def __init__(self):
        super().__init__()
        self.title("Sudoku")

        logger.info("Creating Solution box")
        self.solution = [["" for j in range(9)] for i in range(9)]

        logger.info("Creating String Variables")
        self.sv = [[StringVar() for j in range(9)] for i in range(9)]
        for i in range(9):
            for j in range(9):
                self.sv[i][j].trace(
                    "w", lambda name, index, mode, x=i, y=j: self.detect(x, y)
                )

        logger.info("Creating Frames")
        self.frames = []
        self.headers = ["Sudoku Board", "Actions"]
        for i, header in enumerate(self.headers):
            self.frames.append([])
            self.frames[i].append(LabelFrame(text=header))

        # Creating Sudoku Board
        sudoku_frame = self.frames[0]
        sudoku_frame.append(
            [LabelFrame(sudoku_frame[0], relief="solid") for j in range(9)]
        )
        sudoku_frame.append([])
        for i in range(9):
            sudoku_frame[2].append([])
            for j in range(9):
                row, col = i // 3, j // 3
                k = 3 * row + col
                sudoku_frame[2][i].append(
                    Entry(
                        sudoku_frame[1][k],
                        width=2,
                        font=("Helvetica", 28),
                        justify="center",
                        textvariable=self.sv[i][j],
                    )
                )
                sudoku_frame[2][i][j].bind(
                    "<1>", lambda event, i=i, j=j: self.colour_box(event, i, j)
                )
                sudoku_frame[2][i][j].grid(row=i, column=j)
            sudoku_frame[1][i].grid(row=i // 3, column=i % 3)

        # Creating Actions Buttons
        self.buttons = ["New Game", "Solve", "Hint"]
        button_frame = self.frames[1]
        for j, b in enumerate(self.buttons):
            button_frame.append(
                Button(
                    button_frame[0],
                    text=b,
                    padx=10,
                    pady=5,
                    command=lambda x=b: self.button_action(x),
                )
            )
            button_frame[j + 1].grid(row=1, column=j, padx=38, pady=10)

        # Packing Frames
        for k, frame in enumerate(self.frames):
            frame[0].grid(row=k, column=1, padx=50, pady=10)

        # Generate Board and GUI
        self.generate_board()

    def detect(self, i, j):
        digit = self.sv[i][j].get()

        # Entries must be a digit
        if digit not in DIGITS:
            self.sv[i][j].set("")
        else:
            self.sv[i][j].set(digit)

        # Detect incorrect digit
        if not self.possible(digit, i, j) and digit != "":
            self.frames[0][2][i][j]["bg"] = "#FF7F7F"
        else:
            self.frames[0][2][i][j]["bg"] = "#ADD8E6"

    def colour_box(self, event, i, j):
        # Re-colour selected boxes
        box = 3 * (i // 3) + j // 3
        for row in range(9):
            for col in range(9):
                if row == i and col == j:
                    self.frames[0][2][row][col]["bg"] = "#ADD8E6"
                elif row != i and col == j:
                    self.frames[0][2][row][col]["bg"] = "#daedf4"
                elif row == i and col != j:
                    self.frames[0][2][row][col]["bg"] = "#daedf4"
                elif box == 3 * (row // 3) + col // 3:
                    self.frames[0][2][row][col]["bg"] = "#daedf4"
                else:
                    self.frames[0][2][row][col]["bg"] = "white"
                d = self.frames[0][2][row][col].get()
                if d != "":
                    if not self.possible(d, row, col):
                        self.frames[0][2][row][col]["bg"] = "#FF7F7F"

    def generate_board(self, seed=None, num_clues=17):
        if seed is not None:
            random.seed(seed)
        # Get random positions to keep
        entry_sample = random.sample(range(81), num_clues)
        keep_pos = {(x // 9, x % 9) for x in entry_sample}
        # Refresh the board
        for i in range(9):
            for j in range(9):
                self.sv[i][j].set("")
                self.frames[0][2][i][j]["state"] = "normal"
        # Solve the board
        is_solved = self.solve_board(rand=True)
        # Partially erasing solved board
        for i in range(9):
            for j in range(9):
                self.solution[i][j] = self.sv[i][j].get()
                if (i, j) not in keep_pos:
                    self.sv[i][j].set("")
                else:
                    self.frames[0][2][i][j]["state"] = "disabled"
                self.frames[0][2][i][j]["bg"] = "white"
        if is_solved:
            print("Board Generated.")

    def find_empty(self):
        # Find empty position
        for i in range(9):
            for j in range(9):
                if self.sv[i][j].get() == "":
                    return i, j
        return None

    def possible(self, d, i, j):
        # Check Rows
        for x in range(9):
            if x != i and d == self.sv[x][j].get():
                return False
        # Check Columns
        for y in range(9):
            if y != j and d == self.sv[i][y].get():
                return False
        # Check Box
        for x in range(3 * (i // 3), 3 * (i // 3) + 3):
            for y in range(3 * (j // 3), 3 * (j // 3) + 3):
                if x != i and y != j and d == self.sv[x][y].get():
                    return False
        return True

    def solve_board(self, rand=False):
        # Find Empty positions
        empty_pos = self.find_empty()
        if empty_pos is None:
            return True
        else:
            i, j = empty_pos

        # Try random digits or not
        digits = [str(i) for i in range(1, 10)]
        if rand:
            random.shuffle(digits)

        # Solve via back-tracking
        for d in digits:
            if self.possible(d, i, j):
                self.sv[i][j].set(d)
                if self.solve_board(rand=rand):
                    return True
                self.sv[i][j].set("")
        return False

    def button_action(self, action):
        if action == "New Game":
            self.generate_board()
        elif action == "Solve":
            for i in range(9):
                for j in range(9):
                    if self.sv[i][j].get() == "":
                        self.sv[i][j].set(self.solution[i][j])
        elif action == "Hint":
            entry_sample = random.sample(range(81), 81)
            pos = {(x // 9, x % 9) for x in entry_sample}
            for i, j in pos:
                if self.sv[i][j].get() == "":
                    self.sv[i][j].set(self.solution[i][j])
                    break
        else:
            pass


if __name__ == "__main__":
    logging.basicConfig(level="DEBUG")

    app = SudokuApp()
    app.mainloop()


src/sudoku_app.py
import logging
from tkinter import Button, LabelFrame, Tk

from sudoku_board import SudokuBoard

logger = logging.getLogger(__name__)


class SudokuApp(Tk):
    def __init__(self):
        super().__init__()
        self.title("Sudoku")

        logger.info("Creating Frames")
        self.frames = [
            [LabelFrame(text="Sudoku Board")],
            [LabelFrame(text="Actions")],
        ]

        self.board = SudokuBoard()
        self.actions = LabelFrame(text="Actions")

        # Creating Actions Buttons
        self.buttons = []
        for i, action in enumerate(["New Game", "Solve", "Hint"]):
            self.buttons.append(
                Button(
                    self.actions,
                    text=action,
                    padx=10,
                    pady=5,
                    command=lambda x=action: self.button_action(x),
                )
            )
            self.buttons[i].grid(row=1, column=i, padx=38, pady=10)

        self.actions.grid(row=1, column=1, padx=50, pady=10)

        # Generate Board and GUI
        self.board.generate()

    def button_action(self, action: str):
        if action == "New Game":
            self.board.generate()

        elif action == "Solve":
            self.board.solve()

        elif action == "Hint":
            self.board.hint()

        else:
            pass


if __name__ == "__main__":
    logging.basicConfig(level="DEBUG")

    app = SudokuApp()
    app.mainloop()


src/sudoku/view.py
import logging
from tkinter import Entry, Event, LabelFrame, StringVar
from typing import Optional

from sudoku.logic import SudokuLogic

logger = logging.getLogger(__name__)


class SudokuBoard:
    DIGITS = {str(i) for i in range(1, 11)}

    COLOUR_RIGHT = "#ADD8E6"
    COLOUR_WRONG = "#FF7F7F"

    COLOUR_SELECTED = "#ADD8E6"
    COLOUR_ADJACENT = "#DAEDF4"
    COLOUR_DEFAULT = "#FFFFFF"

    def __init__(self):

        self.logic = SudokuLogic()

        # Create StringVar for each cell on Sudoku Board
        self.sv = [[StringVar() for _ in range(9)] for _ in range(9)]
        for i, j in self.logic.iterate_board():
            self.sv[i][j].trace(
                "w", lambda name, index, mode, x=i, y=j: self.detect(x, y)
            )

        # Create Entry boxes for each cell on Sudoku Board
        self.frame = LabelFrame(text="Sudoku Board")
        self.boxes = [LabelFrame(self.frame, relief="solid") for j in range(9)]
        self.entries: list[list[Entry]] = [[] for _ in range(9)]
        for i in range(9):
            for j in range(9):
                self.entries[i].append(
                    Entry(
                        self.boxes[3 * (i // 3) + (j // 3)],
                        width=2,
                        font=("Helvetica", 28),
                        justify="center",
                        textvariable=self.sv[i][j],
                    )
                )
                self.entries[i][j].bind(
                    "<1>", lambda event, x=i, y=j: self.highlight(event, x, y)
                )
                self.entries[i][j].grid(row=i, column=j)
            self.boxes[i].grid(row=i // 3, column=i % 3)

        # Pack Frames
        self.frame.grid(row=0, column=1, padx=50, pady=10)

    def solve_board(self):
        """Solve the Sudoku board."""
        self.logic.solve_board()
        for i, j in self.logic.iterate_board():
            self.sv[i][j].set(str(self.logic.solution[i][j]))

    def generate(self, seed: Optional[int] = None, num_clues: int = 17):
        keep_pos = self.logic.generate(seed, num_clues)
        # Refresh the board
        for i, j in self.logic.iterate_board():
            self.entries[i][j]["bg"] = "white"
            if (i, j) in keep_pos:
                self.sv[i][j].set(str(self.logic.solution[i][j]))
                self.entries[i][j]["state"] = "disabled"
            else:
                self.sv[i][j].set("")
                self.entries[i][j]["state"] = "normal"
        logger.info("Board Generated.")

    def hint(self):
        """Randomly fill solution for the first empty position on the board."""
        i, j, answer = self.logic.get_hint()
        self.logic.board[i][j] = answer
        self.sv[i][j].set(str(answer))

    def solve(self):
        """Solve the Sudoku board."""
        self.logic.fill_board()
        for i, j in self.logic.iterate_board():
            self.sv[i][j].set(str(self.logic.solution[i][j]))

    def detect(self, i: int, j: int):
        """When String variable in a cell changes, make sure it's a digit."""
        digit = self.sv[i][j].get()

        # Entries must be a digit
        if digit not in self.DIGITS:
            self.sv[i][j].set("")
            self.logic.board[i][j] = 0
            return
        self.sv[i][j].set(digit)
        self.logic.board[i][j] = int(digit)

        # Detect correct digits
        if self.logic.possible(int(digit), i, j):
            self.entries[i][j]["bg"] = self.COLOUR_RIGHT
        else:
            self.entries[i][j]["bg"] = self.COLOUR_WRONG

    def highlight(self, event: Event, i: int, j: int):
        """When an entry is clicked, re-colour cells."""
        logger.debug(f"{event} Event on cell ({i}, {j})")
        for row, col in self.logic.iterate_board():
            # The selected cell
            if row == i and col == j:
                self.entries[row][col]["bg"] = self.COLOUR_SELECTED

            # The cells within row, column, sub-box of selected cell
            elif row != i and col == j:
                self.entries[row][col]["bg"] = self.COLOUR_ADJACENT
            elif row == i and col != j:
                self.entries[row][col]["bg"] = self.COLOUR_ADJACENT
            elif 3 * (i // 3) + j // 3 == 3 * (row // 3) + col // 3:
                self.entries[row][col]["bg"] = self.COLOUR_ADJACENT

            # The remaining cells not within row, column, sub-box of cell
            else:
                self.entries[row][col]["bg"] = self.COLOUR_DEFAULT

            # Colour cells which are invalid
            digit = self.entries[row][col].get()
            if digit != "" and not self.logic.possible(int(digit), row, col):
                self.entries[row][col]["bg"] = self.COLOUR_WRONG


src/sudoku/logic.py
import logging
from copy import deepcopy
from random import sample
from random import seed as random_seed
from random import shuffle
from typing import Generator, Optional

logger = logging.getLogger(__name__)


class SudokuLogic:
    def __init__(self):
        self.empty = [[0 for _ in range(9)] for _ in range(9)]
        self.board: list[list[int]] = deepcopy(self.empty)
        self.solution: list[list[int]] = deepcopy(self.empty)

    @staticmethod
    def iterate_board() -> Generator[tuple[int, int], None, None]:
        """
        Generate all co-ordinates by iterating through the board.

        :return: Yield 2-tuple integer co-ordinates.
        """
        for i in range(9):
            for j in range(9):
                yield i, j

    def find_empty(self) -> Optional[tuple[int, int]]:
        """
        Find the first empty position on the Sudoku board.

        :return: 2-tuple integers co-ordinates of empty position or None.
        """
        for i, j in self.iterate_board():
            if self.board[i][j] == 0:
                return i, j
        return None

    def reset_board(self):
        """Reset the entire board."""
        self.board = deepcopy(self.empty)

    def fill_board(self):
        """Fill the entire board with the solution board."""
        self.board = deepcopy(self.solution)

    def possible(self, digit: int, i: int, j: int) -> bool:
        """
        Determine whether digit can be placed in position (i, j) on the board.

        :param digit: int, the digit to be determined placeable.
        :param i: int, row-coordinate on the Sudoku board.
        :param j: int, column-coordinate on the Sudoku board.
        :return: bool, determines whether digit can be placed.
        """
        # Check Rows
        for x in range(9):
            if x != i and digit == self.board[x][j]:
                return False
        # Check Columns
        for y in range(9):
            if y != j and digit == self.board[i][y]:
                return False
        # Check Box
        for x in range(3 * (i // 3), 3 * (i // 3) + 3):
            for y in range(3 * (j // 3), 3 * (j // 3) + 3):
                if x != i and y != j and digit == self.board[x][y]:
                    return False
        return True

    def solve_board(self) -> bool:
        """
        Solves the Sudoku board using backtracking.

        :return: bool, determines whether the Sudoku board is solved.
        """
        # Find Empty positions
        empty_pos = self.find_empty()
        if empty_pos is None:
            return True
        i, j = empty_pos

        # Try random digits
        digits = [i for i in range(1, 10)]
        shuffle(digits)

        # Solve via backtracking
        for d in digits:
            if self.possible(d, i, j):
                self.board[i][j] = d
                if self.solve_board():
                    return True
                self.board[i][j] = 0
        return False

    def generate(
        self, seed: Optional[int] = None, num_clues: int = 17
    ) -> set[tuple[int, int]]:
        """
        Randomly generate a new board for given seed and number of clues.
        Return the positions of the clues.
        """
        # Set Seed if specified
        if seed is not None:
            random_seed(seed)

        # Get random positions to keep
        keep_pos = {(x // 9, x % 9) for x in sample(range(81), num_clues)}

        # Reset the board
        self.reset_board()

        # Solve the board
        solved = self.solve_board()

        # Partially erase the solved board
        for i, j in self.iterate_board():
            self.solution[i][j] = self.board[i][j]
            if (i, j) not in keep_pos:
                self.board[i][j] = 0
        if solved:
            logger.info("Board Generated.")
        return keep_pos

    def get_hint(self) -> tuple[int, int, int]:
        """Randomly get solution for the first empty position on the board."""
        random_pos = [(x // 9, x % 9) for x in sample(range(81), 81)]
        for i, j in random_pos:
            if self.board[i][j] == 0:
                return i, j, self.solution[i][j]


src/sudoku_board.py
from random import sample
from random import seed as random_seed
from random import shuffle
from tkinter import Entry, Event, LabelFrame, StringVar
from typing import Optional

DIGITS = {str(i) for i in range(1, 11)}


class SudokuBoard:
    def __init__(self):
        self.solution = [["" for _ in range(9)] for _ in range(9)]
        self.sv = [[StringVar() for _ in range(9)] for _ in range(9)]
        for i in range(9):
            for j in range(9):
                self.sv[i][j].trace(
                    "w", lambda name, index, mode, x=i, y=j: self.detect(x, y)
                )
        # Creating Sudoku Board
        self.frame = LabelFrame(text="Sudoku Board")
        self.sub_boxes = [
            LabelFrame(self.frame, relief="solid") for j in range(9)
        ]
        self.board: list[list[Entry]] = [[] for _ in range(9)]
        for i in range(9):
            for j in range(9):
                self.board[i].append(
                    Entry(
                        self.sub_boxes[3 * (i // 3) + (j // 3)],
                        width=2,
                        font=("Helvetica", 28),
                        justify="center",
                        textvariable=self.sv[i][j],
                    )
                )
                self.board[i][j].bind(
                    "<1>", lambda event, x=i, y=j: self.colour_box(event, x, y)
                )
                self.board[i][j].grid(row=i, column=j)
            self.sub_boxes[i].grid(row=i // 3, column=i % 3)

        # Packing Frames
        self.frame.grid(row=0, column=1, padx=50, pady=10)

    def find_empty(self) -> Optional[tuple[int, int]]:
        # Find empty position
        for i in range(9):
            for j in range(9):
                if self.sv[i][j].get() == "":
                    return i, j
        return None

    def solve_board(self, rand: bool = False) -> bool:
        # Find Empty positions
        empty_pos = self.find_empty()
        if empty_pos is None:
            return True
        else:
            i, j = empty_pos

        # Try random digits or not
        digits = [str(i) for i in range(1, 10)]
        if rand:
            shuffle(digits)

        # Solve via back-tracking
        for d in digits:
            if self.possible(d, i, j):
                self.sv[i][j].set(d)
                if self.solve_board(rand=rand):
                    return True
                self.sv[i][j].set("")
        return False

    def generate(self, seed: Optional[int] = None, num_clues: int = 17):
        if seed is not None:
            random_seed(seed)
        # Get random positions to keep
        entry_sample = sample(range(81), num_clues)
        keep_pos = {(x // 9, x % 9) for x in entry_sample}
        # Refresh the board
        for i in range(9):
            for j in range(9):
                self.sv[i][j].set("")
                self.board[i][j]["state"] = "normal"
        # Solve the board
        is_solved = self.solve_board(rand=True)
        # Partially erasing solved board
        for i in range(9):
            for j in range(9):
                self.solution[i][j] = self.sv[i][j].get()
                if (i, j) not in keep_pos:
                    self.sv[i][j].set("")
                else:
                    self.board[i][j]["state"] = "disabled"
                self.board[i][j]["bg"] = "white"
        if is_solved:
            print("Board Generated.")

    def hint(self):
        entry_sample = sample(range(81), 81)
        pos = {(x // 9, x % 9) for x in entry_sample}
        for i, j in pos:
            if self.sv[i][j].get() == "":
                self.sv[i][j].set(self.solution[i][j])
                break

    def solve(self):
        for i in range(9):
            for j in range(9):
                if self.sv[i][j].get() == "":
                    self.sv[i][j].set(self.solution[i][j])

    def detect(self, i: int, j: int):
        digit = self.sv[i][j].get()

        # Entries must be a digit
        if digit not in DIGITS:
            self.sv[i][j].set("")
        else:
            self.sv[i][j].set(digit)

        # Detect incorrect digit
        if not self.possible(digit, i, j) and digit != "":
            self.board[i][j]["bg"] = "#FF7F7F"
        else:
            self.board[i][j]["bg"] = "#ADD8E6"

    def colour_box(self, event: Event, i: int, j: int):
        # Re-colour selected boxes
        box = 3 * (i // 3) + j // 3
        for row in range(9):
            for col in range(9):
                if row == i and col == j:
                    self.board[row][col]["bg"] = "#ADD8E6"
                elif row != i and col == j:
                    self.board[row][col]["bg"] = "#daedf4"
                elif row == i and col != j:
                    self.board[row][col]["bg"] = "#daedf4"
                elif box == 3 * (row // 3) + col // 3:
                    self.board[row][col]["bg"] = "#daedf4"
                else:
                    self.board[row][col]["bg"] = "white"
                digit = self.board[row][col].get()
                if digit != "":
                    if not self.possible(digit, row, col):
                        self.board[row][col]["bg"] = "#FF7F7F"

    def possible(self, digit: str, i: int, j: int) -> bool:
        # Check Rows
        for x in range(9):
            if x != i and digit == self.sv[x][j].get():
                return False
        # Check Columns
        for y in range(9):
            if y != j and digit == self.sv[i][y].get():
                return False
        # Check Box
        for x in range(3 * (i // 3), 3 * (i // 3) + 3):
            for y in range(3 * (j // 3), 3 * (j // 3) + 3):
                if x != i and y != j and digit == self.sv[x][y].get():
                    return False
        return True


src/sudoku/app.py
import logging
from enum import Enum
from tkinter import Button, LabelFrame, Tk

from sudoku.view import SudokuBoard

logger = logging.getLogger(__name__)


class Actions(str, Enum):
    NEW_GAME = "New Game"
    SOLVE = "Solve"
    HINT = "Hint"


class SudokuApp(Tk):
    def __init__(self):
        super().__init__()
        self.title("Sudoku")

        logger.info("Creating Frames")

        self.board = SudokuBoard()
        self.actions = LabelFrame(text="Actions")

        # Creating Actions Buttons
        self.buttons = []
        for i, action in enumerate(Actions):
            self.buttons.append(
                Button(
                    self.actions,
                    text=action.value,
                    padx=10,
                    pady=5,
                    command=lambda x=action: self.button_action(x),
                )
            )
            self.buttons[i].grid(row=1, column=i, padx=38, pady=10)

        self.actions.grid(row=1, column=1, padx=50, pady=10)

        # Generate Board and GUI
        self.board.generate()

    def button_action(self, action: Actions):
        if action == Actions.NEW_GAME:
            self.board.generate(None, 17)

        elif action == Actions.SOLVE:
            self.board.solve()

        elif action == Actions.HINT:
            self.board.hint()

        else:
            pass


if __name__ == "__main__":
    logging.basicConfig(level="INFO")

    app = SudokuApp()
    app.mainloop()
